### PR TITLE
Add new blog post 'Publishing Glean.js'

### PR DESCRIPTION
**Pull request by brizental**
[Rendered](https://github.com/brizental/brizental.github.io/blob/publishing-glean-js/_posts/2021-04-07-publishing-glean-js.markdown)

### CODE REVIEW

File: _posts/2021-04-07-publishing-glean-js.markdown [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+All "This Week in Glean" blog posts are listed in the [TWiG index](https://mozilla.github.io/glean/book/appendix/twig.html)
+(and on the [Mozilla Data blog](https://blog.mozilla.org/data/category/glean/)).
+This article is [cross-posted on the Mozilla Data blog](https://blog.mozilla.org/data/<TODO>).
+
+---
+
+A few weeks ago, it came the time for us to publish the first version of Glean.js in npm. (Yes,
+it has been published. Go [take a look](https://www.npmjs.com/package/@mozilla/glean)). In order
+to publish a package on npm, it is important to define the package entry points in the projects
```
Comment (travis79) [inline _posts/2021-04-07-publishing-glean-js.markdown]:
```suggestion
to publish a package on npm, it is important to define the package entry points in the project's
```

File: _posts/2021-04-07-publishing-glean-js.markdown [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+[`package.json`](https://nodejs.org/api/packages.html) file. The entry point is the path to the file
+that should be loaded when users import a package through `import Package from "package-name"` or
+`const Package = require("package-name")`.
+
+My knowledge in this area, went as far as "Hm, I think that [`main`](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#main)
```
Comment (travis79) [inline _posts/2021-04-07-publishing-glean-js.markdown]:
```suggestion
My knowledge in this area went as far as "Hm, I think that [`main`](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#main)
```

File: _posts/2021-04-07-publishing-glean-js.markdown [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+field in the `package.json` is where we define the entry point, right?". Yes, I was right about that,
+but turns out that was not enough for Glean.js.
```
Comment (travis79) [inline _posts/2021-04-07-publishing-glean-js.markdown]:
```suggestion
but it turns out that was not enough for Glean.js.
```

File: _posts/2021-04-07-publishing-glean-js.markdown [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+## The case of Glean.js
+
+Glean.js is an implementation of Glean for Javascript environments. "Javascript environments" can
+mean multiple things: Node.js servers, Electron apps, websites, webextensions... The list goes on.
+To complicate things, Glean.js needs to access a bunch of platform specific APIs, such as client
```
Comment (travis79) [inline _posts/2021-04-07-publishing-glean-js.markdown]:
```suggestion
To complicate things, Glean.js needs to access a bunch of platform specific APIs such as client
```

File: _posts/2021-04-07-publishing-glean-js.markdown [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  "name": "awesome",
+  "author": "Vandelay Industries",
+  "version": "1.0.0",
+  "main": "./lib/main.js",
+  "types": "./lib/main.d.ts"
+}
+```
+
+Notice the `types` property. It works just like the `main` property. It does not accept an object,
+but only a single entry point. And here we go again, what do you do when the package has multiple
```
Comment (travis79) [inline _posts/2021-04-07-publishing-glean-js.markdown]:
```suggestion
only a single entry point. And here we go again, what do you do when the package has multiple
```

File: _posts/2021-04-07-publishing-glean-js.markdown [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+will need to use bundlers such as [Webpack](https://webpack.js.org/) and [Rollup](https://rollupjs.org/guide/en/),
+which have their own implementation of import/require statement resolutions and are able to import
+both ESM and CommonJS modules either [out-of-the-box](https://webpack.js.org/guides/ecma-script-modules/#flagging-modules-as-esm)
+or through [plugins](https://github.com/rollup/plugins/tree/master/packages/commonjs).
+
+> Note that there is also no need to change the `typesVersions` value for `./node` after this change.
+
+## Final considerations
+
+Although the steps in this post look straight forward enough, it took me quite a while to figure out
```
Comment (travis79) [inline _posts/2021-04-07-publishing-glean-js.markdown]:
```suggestion
Although the steps in this post look straightforward enough, it took me quite a while to figure out
```

File: _posts/2021-04-07-publishing-glean-js.markdown [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+the correct way to configure the Glean.js' entry points. I encountered many caveats along the way, such
+as the `typesVersions` workaround I mentioned above, but also:
+
+- In order to support ES6 modules, it is necessary to include the filename and extension in all internal
+package import statements. CommonJS infers the extension and the filename when it is not provided,
+but ES6 doesn't. This get's extra weird in Glean.js' codebase, because Glean.js is in Typescript
+and still all our import statements have the `.js` extension. See more discussion about this on
```
Comment (travis79) [inline _posts/2021-04-07-publishing-glean-js.markdown]:
```suggestion
and all our import statements still have the `.js` extension. See more discussion about this on
```

File: _posts/2021-04-07-publishing-glean-js.markdown [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+[this issue](https://github.com/microsoft/TypeScript/issues/16577#issuecomment-754941937) and
+[our commit with this change](https://github.com/mozilla/glean.js/pull/123/commits/607c9d5285298f7afbc6187d3b2bdd7d0c1f25b3).
+- Webpack below version 5, does not have support for the `exports` field and is not able to import a
```
Comment (travis79) [inline _posts/2021-04-07-publishing-glean-js.markdown]:
```suggestion
- Webpack, below version 5, does not have support for the `exports` field and is not able to import a
```

File: _posts/2021-04-07-publishing-glean-js.markdown [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+package that defined entry points only using this feature. See the [Webpack 5 release notes](https://webpack.js.org/blog/2020-10-10-webpack-5-release/#major-changes-new-nodejs-ecosystem-features).
+- Other exports conditions such as `browser`, `production` or `development` are mentioned
+in the [Node.js documentation](https://nodejs.org/api/packages.html#packages_conditions_definitions),
+but are ultimately ignored by Node.js. They are used by bundlers such as Webpack and Rollup. The Webpack
+documentation has [a comprehensive list of all the conditions](https://webpack.js.org/guides/package-exports/#conditions)
+you can possibly include in that list, which bundler supports each and whether Node.js supports it too.
```
Comment (travis79) [inline _posts/2021-04-07-publishing-glean-js.markdown]:
```suggestion
you can possibly include in that list, which bundler supports each, and whether Node.js supports it too.
```

File: _posts/2021-04-07-publishing-glean-js.markdown [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+
+Hope this guide is helpful to other people on the internet. Bye! 👋
```
Comment (travis79) [inline _posts/2021-04-07-publishing-glean-js.markdown:194]: Woosh, boy was it ever helpful to me!  I think it was amazing and I learned a ton about the JS implementation of Glean.  Thanks!

**Review by travis79 — APPROVED** [review]
Just a handful of grammar nits, but this was one of the best TWiG's I've read in a while.  🎉

File: _posts/2021-04-07-publishing-glean-js.markdown [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+  "name": "glean",
+  ...
+  "exports": {
+    "./webext": "path/to/entry/point/webext.js",
+    "./node": {
+      "import": "path/to/entry/point/node.js",
+      "require": "path/to/entry/point/node.cjs",
+    },
+    ...
+  }
```
Comment (brizental) [inline _posts/2021-04-07-publishing-glean-js.markdown]:
```suggestion
  },
```

File: _posts/2021-04-07-publishing-glean-js.markdown [hunk 12]
```diff
@@ -0,0 +1,194 @@
+---
+layout: post
+title:  "This Week in Glean: Publishing Glean.js or How I configured a Javascript package that has multiple entry points"
```
Comment (brizental) [inline _posts/2021-04-07-publishing-glean-js.markdown]:
```suggestion
title:  "This Week in Glean: Publishing Glean.js or How I configured an npm package that has multiple entry points"
```

File: _posts/2021-04-07-publishing-glean-js.markdown [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+---
+
+_(“This Week in Glean” is a series of blog posts that the Glean Team at Mozilla is using to try to
+communicate better about our work. They could be release notes, documentation, hopes, dreams, or
+whatever: so long as it is inspired by Glean. You can find
+[an index of all TWiG posts online](https://mozilla.github.io/glean/book/appendix/twig.htm)._
+
+All "This Week in Glean" blog posts are listed in the [TWiG index](https://mozilla.github.io/glean/book/appendix/twig.html)
+(and on the [Mozilla Data blog](https://blog.mozilla.org/data/category/glean/)).
+This article is [cross-posted on the Mozilla Data blog](https://blog.mozilla.org/data/<TODO>).
```
Comment (brizental) [inline _posts/2021-04-07-publishing-glean-js.markdown]:
```suggestion
This article is [cross-posted on the Mozilla Data blog](https://blog.mozilla.org/data/2021/04/07/this-week-in-glean-publishing-glean-js/).
```

File: _posts/2021-04-07-publishing-glean-js.markdown [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+layout: post
+title:  "This Week in Glean: Publishing Glean.js or How I configured a Javascript package that has multiple entry points"
+date:   2021-04-07 18:00:00 +0100
+categories: glean mozilla twig
+---
+
+_(“This Week in Glean” is a series of blog posts that the Glean Team at Mozilla is using to try to
+communicate better about our work. They could be release notes, documentation, hopes, dreams, or
+whatever: so long as it is inspired by Glean. You can find
+[an index of all TWiG posts online](https://mozilla.github.io/glean/book/appendix/twig.htm)._
```
Comment (brizental) [inline _posts/2021-04-07-publishing-glean-js.markdown]:
```suggestion
[an index of all TWiG posts online](https://mozilla.github.io/glean/book/appendix/twig.htm))._
```

**Pull request merged**